flytepropeller/pkg/controller/nodes/array/handler.go

```diff
@@ -631,6 +655,47 @@ func (a *arrayNodeHandler) Handle(ctx context.Context, nCtx interfaces.NodeExecu
 	return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoRunning(&handler.ExecutionInfo{})), nil
 }
 
+// OffloadLargeLiteral offloads the large literal if meets the threshold conditions
+func (a *arrayNodeHandler) OffloadLargeLiteral(ctx context.Context, datastore *storage.DataStore, dataReference storage.DataReference, toBeOffloaded *idlcore.Literal) error {
```

Comment:
Does it make sense to make this a fucntion of the `DataStore`? Or put this in flytestdlib under the storage package somehwere? It seems like we will certainly use it elsewhere as this work progresses.

Reply:
Datastore is working on generic proto message and would want to avoid putting literal functionality over there  as this function does handle infering the literal type and knowing about the offloaded metadata message.

I think if we encounter more use cases we can define a utility in propeller to do this instead of moving to stdlib which  doesn't need to know about the type system IMO. 

Let me know what you think

Reply:
Let me refactor this

Reply:
How does this sound, 


```
// objects
type Options struct {
	// This metadata is passed to stow store implementations.
	Metadata map[string]interface{}
	// Only currently used as an option for protobuf store
	ValidationConfig *ValidationConfig
}

// OffloadingConfig contains configurations for offloading.
// Currently, it is used to configure offloading of large literals.
type ValidationConfig struct {
	MinSizeInMBForOffloading uint64
	MaxSizeInMBForOffloading uint64
}
```

We define a validation Config in store and its used by protbuf store

and the default protobuf implementation does this if the opts contains a validation config


```
func (s DefaultProtobufStore) WriteProtobuf(ctx context.Context, reference DataReference, opts Options, msg proto.Message) error {
	ctx, span := otelutils.NewSpan(ctx, otelutils.BlobstoreClientTracer, "flytestdlib.storage.DefaultProtobufStore/WriteProtobuf")
	defer span.End()

	t := s.metrics.MarshalTime.Start()
	if opts.ValidationConfig != nil {
		literalSizeBytes := uint64(proto.Size(msg))
		literalSizeMB := literalSizeBytes / MB
		// check if the literal is large
		if literalSizeMB > opts.ValidationConfig.MaxSizeInMBForOffloading {
			errWithDetails := fmt.Errorf("message size [%d] MB exceeds the allowed max size [%d] MB", literalSizeMB, opts.ValidationConfig.MaxSizeInMBForOffloading)
			logger.Errorf(ctx, "%v", errWithDetails)
			return errWithDetails
		}
		if literalSizeMB < opts.ValidationConfig.MinSizeInMBForOffloading {
			logger.Debugf(ctx, "message size [%d] MB is smaller than the min size [%d] MB for offloading", literalSizeMB, opts.ValidationConfig.MinSizeInMBForOffloading)
			return nil
		}
	}

```


The caller would just be doing this  and that way we can control how this config is used 

```
opts := storage.Options{
		ValidationConfig: &storage.ValidationConfig{
			MinSizeInMBForOffloading: a.literalOffloadingConfig.MinSizeInMBForOffloading,
			MaxSizeInMBForOffloading: a.literalOffloadingConfig.MaxSizeInMBForOffloading,
		},
	}
	if err := datastore.WriteProtobuf(ctx, dataReference, opts, toBeOffloaded); err != nil {
		logger.Errorf(ctx, "Failed to offload literal at location [%s] with error [%s]", dataReference, err)
		return err
	}
```


A wrapper utlility will take of getting the url and setting it in the literal

```
OffloadLargeLiteral(ctx context.Context, datastore *storage.DataStore, dataReference storage.DataReference, toBeOffloaded *idlcore.Literal) error {
	opts := storage.Options{
		ValidationConfig: &storage.ValidationConfig{
			MinSizeInMBForOffloading: a.literalOffloadingConfig.MinSizeInMBForOffloading,
			MaxSizeInMBForOffloading: a.literalOffloadingConfig.MaxSizeInMBForOffloading,
		},
	}
	if err := datastore.WriteProtobuf(ctx, dataReference, opts, toBeOffloaded); err != nil {
		logger.Errorf(ctx, "Failed to offload literal at location [%s] with error [%s]", dataReference, err)
		return err
	}
	inferredType := validators.LiteralTypeForLiteral(toBeOffloaded)
	if inferredType == nil {
		errString := "Failed to determine literal type for offloaded literal"
		logger.Errorf(ctx, errString)
		return fmt.Errorf(errString)
	}
	// update the literal with the offloaded URI, size and inferred type
	toBeOffloaded.Value = &idlcore.Literal_OffloadedMetadata{
		OffloadedMetadata: &idlcore.LiteralOffloadedMetadata{
			Uri:          dataReference.String(),
			SizeBytes:    uint64(proto.Size(toBeOffloaded)),
			InferredType: inferredType,
		},
	}

	return nil
}
```

Let me know what you think

Reply:
I actually disagree this should live in the datastore since it's more of a logical service layer implementation (we call the datastore only when we choose to offload, right?) and the storage layer is currently agnostic to the message type of the input protobuf

but to Dan's point if we want to use this in copilot down the road or possibly other use cases I quite like having a general purpose helper utility, and not making this function a member of the `arrayNodeHandler` struct

Reply:
Done moved it to utils